File: frontend/src/types/utm.ts
export interface UTMLink {
  id: string
  user_id: string
  title: string | null
  destination_url: string
  short_code: string
  utm_source: string | null
  utm_medium: string | null
  utm_campaign: string | null
  utm_term: string | null
  utm_content: string | null
  created_at: string
  updated_at: string
  full_url: string
  redirect_url: string
  click_count: number
}

export interface UTMLinkCreate {
  title?: string
  destination_url: string
  utm_source?: string
  utm_medium?: string
  utm_campaign?: string
  utm_term?: string
  utm_content?: string
}

export interface UTMLinkListResponse {
  items: UTMLink[]
  total: number
  offset: number
  limit: number
}

export interface ClicksOverTime {
  date: string
  count: number
}

export interface ClicksByDimension {
  label: string
  count: number
}

export interface AnalyticsData {
  clicks_over_time: ClicksOverTime[]
  clicks_by_source: ClicksByDimension[]
  clicks_by_medium: ClicksByDimension[]
  clicks_by_device: ClicksByDimension[]
  clicks_by_browser: ClicksByDimension[]
  clicks_by_country: ClicksByDimension[]
  total_clicks: number
  unique_visitors: number
  total_links: number
  vpn_clicks: number
  days: number
}

export interface LinkAnalyticsData {
  link: UTMLink
  clicks_over_time: ClicksOverTime[]
  clicks_by_device: ClicksByDimension[]
  clicks_by_browser: ClicksByDimension[]
  clicks_by_country: ClicksByDimension[]
  total_clicks: number
  unique_visitors: number
  vpn_clicks: number
  days: number
}

File: backend/app/schemas/utm.py
"""Pydantic schemas for UTM links and analytics.

SOLID-S: Schemas are separate from ORM models.
"""

from __future__ import annotations

from datetime import datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, HttpUrl


class UTMLinkCreate(BaseModel):
    title: Optional[str] = None
    destination_url: str
    utm_source: Optional[str] = None
    utm_medium: Optional[str] = None
    utm_campaign: Optional[str] = None
    utm_term: Optional[str] = None
    utm_content: Optional[str] = None


class UTMLinkResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    user_id: UUID
    title: Optional[str]
    destination_url: str
    short_code: str
    utm_source: Optional[str]
    utm_medium: Optional[str]
    utm_campaign: Optional[str]
    utm_term: Optional[str]
    utm_content: Optional[str]
    created_at: datetime
    updated_at: datetime

    # Computed fields (populated by the service layer)
    full_url: str = ""
    redirect_url: str = ""
    click_count: int = 0


class UTMLinkListResponse(BaseModel):
    items: list[UTMLinkResponse]
    total: int
    offset: int
    limit: int


# ── Analytics ──────────────────────────────────────────────────────────────


class ClicksOverTime(BaseModel):
    date: str
    count: int


class ClicksByDimension(BaseModel):
    label: str
    count: int


class AnalyticsResponse(BaseModel):
    clicks_over_time: list[ClicksOverTime]
    clicks_by_source: list[ClicksByDimension]
    clicks_by_medium: list[ClicksByDimension]
    clicks_by_device: list[ClicksByDimension]
    clicks_by_browser: list[ClicksByDimension]
    clicks_by_country: list[ClicksByDimension]
    total_clicks: int
    unique_visitors: int
    total_links: int
    vpn_clicks: int
    days: int


class LinkAnalyticsResponse(BaseModel):
    link: UTMLinkResponse
    clicks_over_time: list[ClicksOverTime]
    clicks_by_device: list[ClicksByDimension]
    clicks_by_browser: list[ClicksByDimension]
    clicks_by_country: list[ClicksByDimension]
    total_clicks: int
    unique_visitors: int
    vpn_clicks: int
    days: int


class VpnFlagsResponse(BaseModel):
    vpn_clicks: int
    by_country: list[ClicksByDimension]   # VPN clicks grouped by exit-node country
    by_isp: list[ClicksByDimension]       # VPN clicks grouped by ISP/org (provider name)
    days: int
